Adds Ability to Clone a DrakeJoint.

liangfok opened the pull request:
This was extracted out of #4958.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4982)
<!-- Reviewable:end -->

liangfok commented:
+@amcastro-tri for feature review please.

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbpSQX1iRpo6CZgG1Tw:b-57wv7n)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
First round of review. Thank you @liangfok 

---

Reviewed 6 of 16 files at r1.
Review status: 6 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqAqPGWeZPV90VYjOS:-KbqAqPGWeZPV90VYjOT:b9fkhby) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.cc#L128)):*
> ```cpp
>   }
>   for (int i = 0; i < get_num_positions(); ++i) {
>     if (get_position_name(i) != other.get_position_name(i)) {
> ```

You don't need to compare these. This is not comparing actual data but implementation. The result of those methods does not depend on any data, it only depends on the joints having the same type, which you already checked.

---

*[drake/multibody/joints/drake_joint.cc, line 141 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqB2n46C6Tj1MjPDfe:-KbqB2n46C6Tj1MjPDff:bbxgkzp) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.cc#L141)):*
> ```cpp
>   }
>   for (int i = 0; i < get_num_velocities(); ++i) {
>     if (get_velocity_name(i) != other.get_velocity_name(i)) {
> ```

Same comment as above

---

*[drake/multibody/joints/drake_joint.cc, line 153 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqB68g0r4VQNzU7YwQ:-KbqB68g0r4VQNzU7YwR:blzjxlf) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.cc#L153)):*
> ```cpp
>     }
>   }
>   if (is_floating() != other.is_floating()) {
> ```

You already checked the joint type, this is not needed. You only need to compare actual in-memory data. 
Take this as an example: If you compare these, then would that mean that you also have to compare the output result for any other method in a `DrakeJoint` to verify the joints are equivalent? clearly not.

---

*[drake/multibody/joints/drake_joint.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBYFWBFYgA-7n3vY5:-KbqBYFWBFYgA-7n3vY6:b-vheu38) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.cc#L171)):*
> ```cpp
>     return false;
>   }
>   if (zeroConfiguration() != other.zeroConfiguration()) {
> ```

Same comment, you are comparing method results, not in-memory data.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-Kbq7LZxuGVoMx31_4v4:b-p9bowk) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
> ```objc
>   EIGEN_MAKE_ALIGNED_OPERATOR_NEW
> 
>   virtual bool operator==(const DrakeJoint& other) const;
> ```

Making `operator==()` usually does not make sense (I don't even know in what case it would). The reason is that the `operator==` in the derived class will have a different signature.
I see however that you are doing things like `RevoluteJoint::opeator==(const DrakeJoint& other)`. That actually is a bad idea. Think about what it really means. It means that you can actually compare a `RevoluteJoint` to any other `DrakeJoint`. That is not what you want with an `operator==()`.
Also, notice that many of the comparisons are actually floating point arithmetic comparisons and then a "near" comparison is more appropriate.
I propose then to write a method called something like `IsApproximatelyEqual(const DrakeJoint& other, double tolerance = NumTraits<double>::epsilon())` that uses Eigen's `isApprox()` method and remove `operator==()` overloads completely. Use `IsApproximatelyEqual()` instead.

---

*[drake/multibody/joints/fixed_axis_one_dof_joint.h, line 204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq97CfKJMXmN8UznSJ:-Kbq97CgMvW-Vm7zyT-d:b-lb043v) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/fixed_axis_one_dof_joint.h#L204)):*
> ```objc
>       return false;
>     }
>     if (joint_axis != downcasted_joint->joint_axis) {
> ```

Here a floating point comparison is more appropriate `joint_axis.isApprox(downcasted_joint->joint_axis, tolerance)`

---

*[drake/multibody/joints/helical_joint.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBpiJ0VM2bHxYKdaV:-KbqBpiJ0VM2bHxYKdaW:bt7qjfd) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/helical_joint.cc#L34)):*
> ```cpp
>     return false;
>   }
>   if (axis_ != downcasted_joint->axis_) {
> ```

Use `isApprox()`

---

*[drake/multibody/joints/helical_joint.cc, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBtVmqnPYRWV7d9id:-KbqBtVmqnPYRWV7d9ie:b-t0e955) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/helical_joint.cc#L43)):*
> ```cpp
>     return false;
>   }
>   if (pitch_ != downcasted_joint->pitch_) {
> ```

use `abs(pitch_ - other.pitch) < epsilon`

---

*[drake/multibody/joints/test/joint_test.cc, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqCEyxVfWEVzZx1cMe:-KbqCEyxVfWEVzZx1cMf:brym7t3) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/test/joint_test.cc#L65)):*
> ```cpp
> TEST_F(DrakeJointTests, TestCloneAndCompare) {
>   // Verifies that a joint is equal to itself.
>   EXPECT_EQ(*fixed_joint_, *fixed_joint_);
> ```

This tests here are exactly the same as the one below with the `for` loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbqCKyUsmxHpHyF5Byb:baqy0ce)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-KbqCdYOLMGk9E0rbz6w:b7caf6z) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Making `operator==()` usually does not make sense (I don't even know in what case it would). The reason is that the `operator==` in the derived class will have a different signature.
I see however that you are doing things like `RevoluteJoint::opeator==(const DrakeJoint& other)`. That actually is a bad idea. Think about what it really means. It means that you can actually compare a `RevoluteJoint` to any other `DrakeJoint`. That is not what you want with an `operator==()`.
Also, notice that many of the comparisons are actually floating point arithmetic comparisons and then a "near" comparison is more appropriate.
I propose then to write a method called something like `IsApproximatelyEqual(const DrakeJoint& other, double tolerance = NumTraits<double>::epsilon())` that uses Eigen's `isApprox()` method and remove `operator==()` overloads completely. Use `IsApproximatelyEqual()` instead.


</blockquote></details>

Correction: Making `operator==()` **virtual** usually does not make sense ...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 6 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-KbqDyWoVEJRpvNJtmcn:bo38y47) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
> It means that you can actually compare a RevoluteJoint to any other DrakeJoint. That is not what you want with an operator==().

In `RevoluteJoint`'s implementation of `operator==`, it attempts to downcast the `DrakeJoint` into a `RevoluteJoint` and returns false if the downcast is unsuccessful. Isn't this the desired behavior?

> Also, notice that many of the comparisons are actually floating point arithmetic comparisons and then a "near" comparison is more appropriate.

In general I would agree, but in this context, `operator==` is being used to test a `Clone()` operation, which should return a joint with exactly the same floating point values. How about we use Eigen's `isApprox()` or Drake's `is_approx_equal_abstol()` within the `operator==` method?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-KbqnOYDA0fAMCXRbk_J:bdvn2kk) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

> It means that you can actually compare a RevoluteJoint to any other DrakeJoint. That is not what you want with an operator==().

In `RevoluteJoint`'s implementation of `operator==`, it attempts to downcast the `DrakeJoint` into a `RevoluteJoint` and returns false if the downcast is unsuccessful. Isn't this the desired behavior?

> Also, notice that many of the comparisons are actually floating point arithmetic comparisons and then a "near" comparison is more appropriate.

In general I would agree, but in this context, `operator==` is being used to test a `Clone()` operation, which should return a joint with exactly the same floating point values. How about we use Eigen's `isApprox()` or Drake's `is_approx_equal_abstol()` within the `operator==` method?
</blockquote></details>

I agree with @amcastro-tri that this is not a good use of operator==(). Instead, just make a clearly-named method that performs the operation you want to perform. There are too many possible meanings for '=='.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-Kbtxw_B_-1NVbRZOKiT:bcitz4z) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I agree with @amcastro-tri that this is not a good use of operator==(). Instead, just make a clearly-named method that performs the operation you want to perform. There are too many possible meanings for '=='.
</blockquote></details>

One second thought here @liangfok, sorry for the confusion. I still do feel strong about the `operator==()` here, it is not a good use in this way. As @sherm1 mentioned, a clearly named method (you can make it virtual of course) like `IsExactlyEqualTo(const DrakeJoint& other)` should get the job done.

However, I am going to correct myself on the `isApprox()` idea. It is a bad idea, sorry. As you said, since you are testing the result of a `Clone()` operation, we should expect the two joints to be equal byte by byte (That is why above now I suggested the `IsExactlyEqualTo()` method name). Maybe even the compare method could be called `CompareToClonedJoint(const DrakeJoint& other)` since really this kind of byte-to-byte comparison would only be useful for that. Hopefully this clarifies.

---

*[drake/multibody/joints/fixed_axis_one_dof_joint.h, line 204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq97CfKJMXmN8UznSJ:-Kbtz1NAr--aBc3iHse5:bfgp4av) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/fixed_axis_one_dof_joint.h#L204)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Here a floating point comparison is more appropriate `joint_axis.isApprox(downcasted_joint->joint_axis, tolerance)`
</blockquote></details>

Disregard comment

---

*[drake/multibody/joints/helical_joint.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBpiJ0VM2bHxYKdaV:-Kbtz539ZvylAndBML_k:bnkl1oo) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/helical_joint.cc#L34)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Use `isApprox()`
</blockquote></details>

Disregard

---

*[drake/multibody/joints/helical_joint.cc, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBtVmqnPYRWV7d9id:-Kbtz7u8Vqm_cw4NZjUD:bnkl1oo) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/helical_joint.cc#L43)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

use `abs(pitch_ - other.pitch) < epsilon`
</blockquote></details>

Disregard

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 6 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-Kbu1msEfQjBaJfb16Qz:b-k0b3kg) ([raw file](https://github.com/robotlocomotion/drake/blob/ec3c3dd9cd0d5f6ee8938412676b5c3762819247/drake/multibody/joints/drake_joint.h#L229)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

One second thought here @liangfok, sorry for the confusion. I still do feel strong about the `operator==()` here, it is not a good use in this way. As @sherm1 mentioned, a clearly named method (you can make it virtual of course) like `IsExactlyEqualTo(const DrakeJoint& other)` should get the job done.

However, I am going to correct myself on the `isApprox()` idea. It is a bad idea, sorry. As you said, since you are testing the result of a `Clone()` operation, we should expect the two joints to be equal byte by byte (That is why above now I suggested the `IsExactlyEqualTo()` method name). Maybe even the compare method could be called `CompareToClonedJoint(const DrakeJoint& other)` since really this kind of byte-to-byte comparison would only be useful for that. Hopefully this clarifies.
</blockquote></details>

Alright, I agree with the above arguments and will adjust this PR as recommended. Thanks for the review comments!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 1 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqAqPGWeZPV90VYjOS:-Kbu6Mc1jGsreo45-FmF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/drake_joint.cc#L128)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

You don't need to compare these. This is not comparing actual data but implementation. The result of those methods does not depend on any data, it only depends on the joints having the same type, which you already checked.
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.cc, line 141 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqB2n46C6Tj1MjPDfe:-Kbu6NYFeIekCqXBdG6V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/drake_joint.cc#L141)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Same comment as above
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.cc, line 153 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqB68g0r4VQNzU7YwQ:-Kbu6Rn3m1I4rfiy7GC4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/drake_joint.cc#L153)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

You already checked the joint type, this is not needed. You only need to compare actual in-memory data. 
Take this as an example: If you compare these, then would that mean that you also have to compare the output result for any other method in a `DrakeJoint` to verify the joints are equivalent? clearly not.
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqBYFWBFYgA-7n3vY5:-Kbu6W0Zh3VhGWgx0ic0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/drake_joint.cc#L171)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Same comment, you are comparing method results, not in-memory data.
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.h, line 229 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq7LZxuGVoMx31_4v3:-KbuKj8IbIjd0kvALXms:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/drake_joint.h#L229)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alright, I agree with the above arguments and will adjust this PR as recommended. Thanks for the review comments!
</blockquote></details>

Done.

---

*[drake/multibody/joints/fixed_axis_one_dof_joint.h, line 204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbq97CfKJMXmN8UznSJ:-Kbu6Zu-GQJBHJXP-ipm:b-lsh891) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/fixed_axis_one_dof_joint.h#L204)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Disregard comment
</blockquote></details>

OK. No Problem.

---

*[drake/multibody/joints/test/joint_test.cc, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbqCEyxVfWEVzZx1cMe:-KbuBkc-ud8qzHP3wwJa:bxdc7fw) ([raw file](https://github.com/robotlocomotion/drake/blob/c391b386caedfe0d16d09e169e0a067d9c51cd43/drake/multibody/joints/test/joint_test.cc#L65)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This tests here are exactly the same as the one below with the `for` loop.
</blockquote></details>

Done. (Removed.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbuBlt3Jh4LomY3wbo6:bblnx7i)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 5 of 16 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuqqIH5B_obXDup4zX:-KbuqqIH5B_obXDup4zY:b-x6w54a) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L89)):*
> ```cpp
> }
> 
> bool DrakeJoint::CompareToClonedJoint(const DrakeJoint &other) const {
> ```

BTW, consider using RTTI here to avoid repeating for each derived class the type check. RTTI also gives you a string name you can use in your log message.

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-Kbujmx1D0NpjVFUdHW-:ble5d5a) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
> ```cpp
>   if (get_transform_to_parent_body().matrix() !=
>       other.get_transform_to_parent_body().matrix()) {
>     drake::log()->debug(
> ```

BTW, do you know if there is a way to unit test `drake::log()`'s?

---

*[drake/multibody/joints/drake_joint.cc, line 128 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbukGLEYABHX95qSoQ8:-KbukGLEYABHX95qSoQ9:b-nu5v8w) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L128)):*
> ```cpp
>     return false;
>   }
>   if (is_fixed() != other.is_fixed()) {
> ```

not needed. You already compared `num_positions`

---

*[drake/multibody/joints/drake_joint.h, line 231 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbumz_uz2TCoOerZCGV:-Kbumz_uz2TCoOerZCGW:b4geb12) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.h#L231)):*
> ```objc
>   // Compares this joint with a cloned joint. Since this method is intended to
>   // compare a clone, an *exact* match is performed. This method will only
>   // return true if the provided `other` joint is exactly the same as this
> ```

`true` (between single quotes.

---

*[drake/multibody/joints/drake_joint.h, line 232 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbumwX2oYJJXYwz44F6:-KbumwX2oYJJXYwz44F7:bcz5tu6) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.h#L232)):*
> ```objc
>   // compare a clone, an *exact* match is performed. This method will only
>   // return true if the provided `other` joint is exactly the same as this
>   // joint.
> ```

Use `///` for Doxygen.

---

*[drake/multibody/joints/fixed_joint.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunnZx6lqdDvXmzKbX:-KbunnZx6lqdDvXmzKbY:b2wemhx) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/fixed_joint.h#L19)):*
> ```objc
>   virtual ~FixedJoint() {}
> 
>   std::unique_ptr<DrakeJoint> Clone() const override;
> ```

`final`

---

*[drake/multibody/joints/helical_joint.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunsCOZXpZt5Duq0Em:-KbunsCOZXpZt5Duq0En:b2wemhx) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/helical_joint.h#L24)):*
> ```objc
>   virtual ~HelicalJoint() {}
> 
>   std::unique_ptr<DrakeJoint> Clone() const override;
> ```

`final`

---

*[drake/multibody/joints/prismatic_joint.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuobozZCTtGLdvINyG:-KbuobozZCTtGLdvINyH:b5qpbfw) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.cc#L25)):*
> ```cpp
> bool PrismaticJoint::CompareToClonedJoint(const DrakeJoint& other) const {
>   if (!FixedAxisOneDoFJoint::CompareToClonedJoint(other)) return false;
>   const PrismaticJoint* downcasted_joint =
> ```

BTW, There is a lot of repeated code here (I mean the part where you compare the two dynamic casts and print a message). Consider using RTTI. I think that would be a use case of RTTI here.
With RTTI you could check the type in `DrakeJoint::CompareToClonedJoint()`

---

*[drake/multibody/joints/prismatic_joint.h, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunwA3tJikD5k-dM8O:-KbunwA3tJikD5k-dM8P:b2wemhx) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.h#L51)):*
> ```objc
>   virtual ~PrismaticJoint() {}
> 
>   std::unique_ptr<DrakeJoint> Clone() const override;
> ```

`final`

---

*[drake/multibody/joints/prismatic_joint.h, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunxUsF3zem4zcKxnu:-KbunxUsF3zem4zcKxnv:bks2pci) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.h#L53)):*
> ```objc
>   std::unique_ptr<DrakeJoint> Clone() const override;
> 
>   bool CompareToClonedJoint(const DrakeJoint& other) const override;
> ```

`final`, well, ok, you get the idea :-)

---

*[drake/multibody/joints/test/joint_test.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiMVEywfhZAmoDw3x:-KbuiMVEywfhZAmoDw3y:bxml9ny) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L75)):*
> ```cpp
>   // Verifies that two joints with different names do not match.
>   EXPECT_FALSE(fixed_joint_->CompareToClonedJoint(
>             FixedJoint("NonMatchingName",
> ```

strange indentation.

---

*[drake/multibody/joints/test/joint_test.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiOC83jHdLI5UaKoj:-KbuiOC83jHdLI5UaKok:b-ahk7g1) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L89)):*
> ```cpp
>   Eigen::Isometry3d non_matching_transform =
>       joints.at(0)->get_transform_to_parent_body();
>   non_matching_transform.linear()(0, 0) += 1;
> ```

BTW, the transform here created actually is not a valid (physical) transformation. Probably a better option would be to use `AngleAxis` or similar mechanism.

---

*[drake/multibody/joints/test/joint_test.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiByKbsBl2PYvAtPN:-KbuiByLZsvAcir5zMgI:b-n0di04) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L91)):*
> ```cpp
>   non_matching_transform.linear()(0, 0) += 1;
>   EXPECT_FALSE(joints.at(0)->CompareToClonedJoint(
>             FixedJoint(fixed_joint_->get_name(), non_matching_transform)));
> ```

strange indentation.

---

*[drake/multibody/joints/test/joint_test.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuimS_nBOFMEReZ5MJ:-KbuimS_nBOFMEReZ5MK:b-dge1i9) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L96)):*
> ```cpp
>   // the original.
>   for (int i = 0; i < static_cast<int>(joints.size()); ++i) {
>     EXPECT_TRUE(joints.at(i)->CompareToClonedJoint(*joints.at(i)->Clone()));
> ```

BYW, just curious, any reason you use `at()` instead of `operator[]`?

---

*[drake/multibody/joints/test/joint_test.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbuj5A_luao6VRbEtQI:-Kbuj5AaI5NNQzf-H3s6:b-upgl5r) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L99)):*
> ```cpp
>   }
> 
>   // Verifies that two joints of different types do not match.
> ```

BTW, you are also verifying that `joint1->CompareToClonedJoint(*joint2) == joint2->CompareToClonedJoint(*joint1)` since the loops exhaust the pair `i, j` as well as the pair `j, i`. i.e. this verifies your implementation is commutative, a good thing!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 7 of 16 files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuqqIH5B_obXDup4zX:-Kbuu6ZUXlIoLnpSEnC0:b-x5l7mh) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, consider using RTTI here to avoid repeating for each derived class the type check. RTTI also gives you a string name you can use in your log message.
</blockquote></details>

I considered adding the following logic:

```
#include <typeinfo>
...
bool DrakeJoint::CompareToClonedJoint(const DrakeJoint &other) const {
  if (typeid(*this) != typeid(other)) {
    drake::log()->debug("DrakeJoint::CompareToClonedJoint(): "
        "type mismatch:\n"
        "  - this: {}\n"
        "  - other: {}",
        typeid(*this),
        typeid(other));
    return false;
  }
  ...
}
```

But I decided against it since the sub-classes need a pointer to the down-casted joint anyway and it's good practice to always verify that the pointer to the down-casted joint is not `nullptr`. Then again, I suppose it wouldn't hurt to include the above logic just to be even more safe.

Thoughts?

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-Kburpy_3KjgM68Mcgc1:boe22f0) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, do you know if there is a way to unit test `drake::log()`'s?
</blockquote></details>

Yes, there's this:

http://stackoverflow.com/questions/3803465/how-to-capture-stdout-stderr-with-googletest

Would you like the unit test to to cover these log statements?

---

*[drake/multibody/joints/drake_joint.cc, line 128 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbukGLEYABHX95qSoQ8:-Kbuv7IJlZtaVx0I7qBK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L128)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

not needed. You already compared `num_positions`
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.h, line 231 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbumz_uz2TCoOerZCGV:-KbuvK2l0yKbBUFbnQ_A:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.h#L231)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`true` (between single quotes.
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.h, line 232 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbumwX2oYJJXYwz44F6:-KbuvKdSMKN2D___sAb9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.h#L232)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Use `///` for Doxygen.
</blockquote></details>

Done.

---

*[drake/multibody/joints/fixed_joint.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunnZx6lqdDvXmzKbX:-KbuvZrni4CCMzaxKCJW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/fixed_joint.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`final`
</blockquote></details>

Done.

---

*[drake/multibody/joints/helical_joint.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunsCOZXpZt5Duq0Em:-Kbuvc5j_Wr3SvvZ8x9b:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/helical_joint.h#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`final`
</blockquote></details>

Done.

---

*[drake/multibody/joints/prismatic_joint.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuobozZCTtGLdvINyG:-KbuvjWTkefstxAtRhB8:bww1x3c) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.cc#L25)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, There is a lot of repeated code here (I mean the part where you compare the two dynamic casts and print a message). Consider using RTTI. I think that would be a use case of RTTI here.
With RTTI you could check the type in `DrakeJoint::CompareToClonedJoint()`
</blockquote></details>

See my comment in `drake_joint.cc`. Performing RTTI there is possible, but I'll still need to have the down-casting logic here and might as well check that the resulting pointer is valid.

---

*[drake/multibody/joints/prismatic_joint.h, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunwA3tJikD5k-dM8O:-Kbuw0Dd8P7Sg5CW9f0g:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.h#L51)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`final`
</blockquote></details>

Done.

---

*[drake/multibody/joints/prismatic_joint.h, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbunxUsF3zem4zcKxnu:-Kbuw9fyfVe-63xWYP0N:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.h#L53)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`final`, well, ok, you get the idea :-)
</blockquote></details>

Done.

---

*[drake/multibody/joints/test/joint_test.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiMVEywfhZAmoDw3x:-KbuwHJqY94HaUueJXjC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L75)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

strange indentation.
</blockquote></details>

Done.

---

*[drake/multibody/joints/test/joint_test.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiOC83jHdLI5UaKoj:-KbuxkjMCbom0KOaXS1M:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, the transform here created actually is not a valid (physical) transformation. Probably a better option would be to use `AngleAxis` or similar mechanism.
</blockquote></details>

Done.

---

*[drake/multibody/joints/test/joint_test.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuiByKbsBl2PYvAtPN:-KbuxsnSAI7_Huc0F6X2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L91)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

strange indentation.
</blockquote></details>

Done.

---

*[drake/multibody/joints/test/joint_test.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuimS_nBOFMEReZ5MJ:-Kbuxv9PCR-LxCtwxk0n:b6k5oo) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BYW, just curious, any reason you use `at()` instead of `operator[]`?
</blockquote></details>

I believe `at()` is safer than `operator[]` because it performs bounds checking.

---

*[drake/multibody/joints/test/joint_test.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbuj5A_luao6VRbEtQI:-KbuyAoqWt8qAA46INeB:bwuv9uq) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L99)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, you are also verifying that `joint1->CompareToClonedJoint(*joint2) == joint2->CompareToClonedJoint(*joint1)` since the loops exhaust the pair `i, j` as well as the pair `j, i`. i.e. this verifies your implementation is commutative, a good thing!
</blockquote></details>

Nice!!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-Kbv1RltBA5NAEUF4f4A:bblnx7i)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I think this is ready for platform @liangfok. 
+@sammy-tri please since it is your platform review day.

---

Reviewed 7 of 13 files at r3.
Review status: 10 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuqqIH5B_obXDup4zX:-Kbv8pAOUgSBwqrp_7Qs:b6xu20j) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L89)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I considered adding the following logic:

```
#include <typeinfo>
...
bool DrakeJoint::CompareToClonedJoint(const DrakeJoint &other) const {
  if (typeid(*this) != typeid(other)) {
    drake::log()->debug("DrakeJoint::CompareToClonedJoint(): "
        "type mismatch:\n"
        "  - this: {}\n"
        "  - other: {}",
        typeid(*this),
        typeid(other));
    return false;
  }
  ...
}
```

But I decided against it since the sub-classes need a pointer to the down-casted joint anyway and it's good practice to always verify that the pointer to the down-casted joint is not `nullptr`. Then again, I suppose it wouldn't hurt to include the above logic just to be even more safe.

Thoughts?
</blockquote></details>

Maybe you can have a `template <class DowcastToType> DowncastToType* DrakeJoint::DowncastOrLog(const DrakeJoint* other)` that implements the above logic and returns the result of `dynamic_cast<DowncastToType*>`?

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-Kbv9_18n6m8yCj_QgSH:b-pf68ke) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, there's this:

http://stackoverflow.com/questions/3803465/how-to-capture-stdout-stderr-with-googletest

Would you like the unit test to to cover these log statements?
</blockquote></details>

Not really, just being curious. Also those statements happen only on `Debug` builds right?

---

*[drake/multibody/joints/prismatic_joint.cc, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvA4wq0WzBTTmNdHvs:-KbvA4wq0WzBTTmNdHvt:byiqz32) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.cc#L24)):*
> ```cpp
> 
> bool PrismaticJoint::CompareToClonedJoint(const DrakeJoint& other) const {
>   if (!FixedAxisOneDoFJoint::CompareToClonedJoint(other)) return false;
> ```

BTW, With my suggestion in `drake_joint.cc` this line would read like:
```c++
if (!(downcasted_joint = DrakeJoint::DowncastOrLog<PrismaticJoint>(&other))) return false;
```

maybe the platform reviewer will have a better suggestion.

---

*[drake/multibody/joints/test/joint_test.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuimS_nBOFMEReZ5MJ:-KbvBRLI7WMGA3jS-PVS:b-ij92x3) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L96)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe `at()` is safer than `operator[]` because it performs bounds checking.
</blockquote></details>

Yes, I know. Is that the reason here. Also because of that `.at()` is slightly slower, something to keep in mind if you want fast code.
Anyway, I thought you'd have a some sort of either google style guide reason or gtest framework reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbvBybqzYWXxXhLnc1x:b-j476h5)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 6 of 13 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 12 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuqqIH5B_obXDup4zX:-KbvG2sj-VvmO29KnfS4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Maybe you can have a `template <class DowcastToType> DowncastToType* DrakeJoint::DowncastOrLog(const DrakeJoint* other)` that implements the above logic and returns the result of `dynamic_cast<DowncastToType*>`?
</blockquote></details>

Done.

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-KbvD_1cuVL96zVp6x6j:b-yrkzvt) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Not really, just being curious. Also those statements happen only on `Debug` builds right?
</blockquote></details>

Correct. Here is the logic:

https://github.com/RobotLocomotion/drake/blob/6a7beb5876a75c1a7522589e49e5b8195668bc0e/drake/common/text_logging.h#L33

---

*[drake/multibody/joints/prismatic_joint.cc, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvA4wq0WzBTTmNdHvs:-KbvHz15ptBmW0npxyuv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/prismatic_joint.cc#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, With my suggestion in `drake_joint.cc` this line would read like:
```c++
if (!(downcasted_joint = DrakeJoint::DowncastOrLog<PrismaticJoint>(&other))) return false;
```

maybe the platform reviewer will have a better suggestion.
</blockquote></details>

Done.

---

*[drake/multibody/joints/test/joint_test.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuimS_nBOFMEReZ5MJ:-KbvHcCehqdmAN-55MvW:b-lxlus4) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Yes, I know. Is that the reason here. Also because of that `.at()` is slightly slower, something to keep in mind if you want fast code.
Anyway, I thought you'd have a some sort of either google style guide reason or gtest framework reason.
</blockquote></details>

Ah, I see. There's no other reason behind my selection of `at()` over `operator[]`. Thanks for the advice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbvHqgmmvVW6tQFCsDZ:bblnx7i)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 3 of 16 files at r1, 13 of 13 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 234 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvL7lWt7yZJVSTphiK:-KbvL7lWt7yZJVSTphiL:bpfzgle) ([raw file](https://github.com/robotlocomotion/drake/blob/e653707eab8d3639850da106bb71630e4211b2cb/drake/multibody/joints/drake_joint.h#L234)):*
> ```objc
>   /// Otherwise, it returns `nullptr`.
>   template <class DowncastType>
>   const DowncastType* DowncastOrLog(const DrakeJoint* other) const {
> ```

BTW could this be made protected?

---

*[drake/multibody/joints/quaternion_floating_joint.cc, line 15 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvKOPg3-8TQS4TRq-z:-KbvKOPhc3HnCerYnUAc:b1qfojj) ([raw file](https://github.com/robotlocomotion/drake/blob/39d7f038b7e9cb9bb4291d0bf9279f4053835632/drake/multibody/joints/quaternion_floating_joint.cc#L15)):*
> ```cpp
> std::unique_ptr<DrakeJoint> QuaternionFloatingJoint::Clone() const {
>   auto joint = std::make_unique<QuaternionFloatingJoint>(get_name(),
>                                                 get_transform_to_parent_body());
> ```

This indentation looks odd...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 14 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.h, line 234 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvL7lWt7yZJVSTphiK:-KbvLrcjR7Mgmaq73ZF2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e653707eab8d3639850da106bb71630e4211b2cb/drake/multibody/joints/drake_joint.h#L234)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW could this be made protected?
</blockquote></details>

Done.

---

*[drake/multibody/joints/quaternion_floating_joint.cc, line 15 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvKOPg3-8TQS4TRq-z:-KbvM0QIORwOMVS1YQwZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39d7f038b7e9cb9bb4291d0bf9279f4053835632/drake/multibody/joints/quaternion_floating_joint.cc#L15)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This indentation looks odd...
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you @liangfok! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 3 of 4 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-KbvP2pQ4MKWeEQ3Vdo4:b-6j2j57) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Correct. Here is the logic:

https://github.com/RobotLocomotion/drake/blob/6a7beb5876a75c1a7522589e49e5b8195668bc0e/drake/common/text_logging.h#L33

</blockquote></details>

Great!. But that would make writing a unit test more difficult right? I mean, it only needs to run for Debug builds? again just for my own education, I won't ask you to do that here.

---

*[drake/multibody/joints/drake_joint.h, line 234 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvL7lWt7yZJVSTphiK:-KbvQEihy3mES9wZr6a1:b-qluu3g) ([raw file](https://github.com/robotlocomotion/drake/blob/e653707eab8d3639850da106bb71630e4211b2cb/drake/multibody/joints/drake_joint.h#L234)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW, Nice observation!

---

*[drake/multibody/joints/drake_joint.h, line 239 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvQI0QvL_LTqVq-Nk7:-KbvQI0QvL_LTqVq-Nk8:bxhnwm1) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/drake_joint.h#L239)):*
> ```objc
>   /// Attempts to downcast the provided `other` to the template class type. If
>   /// the downcast is successful, it returns a pointer to the downcasted type.
>   /// Otherwise, it returns `nullptr`.
> ```

Maybe comment here that this will log to screen on failure before returing?
Also, now this can be made a normal comment since it is not public anymore.

---

*[drake/multibody/joints/helical_joint.cc, line 29 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvQao9hKlADTBAI9NM:-KbvQao9hKlADTBAI9NN:bxbef0u) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/helical_joint.cc#L29)):*
> ```cpp
>   const HelicalJoint* downcasted_joint =
>       DrakeJoint::DowncastOrLog<HelicalJoint>(&other);
>   if (downcasted_joint == nullptr) {
> ```

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.

---

*[drake/multibody/joints/prismatic_joint.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvPbUBBIhAAjPLWY5Z:-KbvPbUBBIhAAjPLWY5_:b-wn9vwk) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/prismatic_joint.cc#L27)):*
> ```cpp
>   const PrismaticJoint* downcasted_joint =
>       DrakeJoint::DowncastOrLog<PrismaticJoint>(&other);
>   if (downcasted_joint == nullptr) {
> ```

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.

---

*[drake/multibody/joints/revolute_joint.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvPkxjf46Qz061SOGZ:-KbvPkxjf46Qz061SOG_:bxbef0u) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/revolute_joint.cc#L27)):*
> ```cpp
>   const RevoluteJoint* downcasted_joint =
>       DrakeJoint::DowncastOrLog<RevoluteJoint>(&other);
>   if (downcasted_joint == nullptr) {
> ```

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.

---

*[drake/multibody/joints/test/joint_test.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbuimS_nBOFMEReZ5MJ:-KbvPq3hv25DwOCjG8av:beujrf9) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/test/joint_test.cc#L96)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Ah, I see. There's no other reason behind my selection of `at()` over `operator[]`. Thanks for the advice!
</blockquote></details>

No advice really, I was trying to learn why or the other. Not that I'm strong about the usage of `at()` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbvQiY7r7wYz634GTrn:b-lqt2no)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 13 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/joints/drake_joint.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-Kbujmx1D0NpjVFUdHVz:-KbvRP-j1_9gYkjhNSb3:b-ycmg3k) ([raw file](https://github.com/robotlocomotion/drake/blob/39cc5903ad7ab412158afa51b06b6b05a956f22f/drake/multibody/joints/drake_joint.cc#L92)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Great!. But that would make writing a unit test more difficult right? I mean, it only needs to run for Debug builds? again just for my own education, I won't ask you to do that here.
</blockquote></details>

Yes, it would require modifying the unit test to only run on debug builds. Maybe something like:

```
#ifndef NDEBUG
/* Debug only unit test */    
#endif 
```

---

*[drake/multibody/joints/drake_joint.h, line 239 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvQI0QvL_LTqVq-Nk7:-KbvS_tAbgCsRoSkXFMM:bf343kf) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/drake_joint.h#L239)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Maybe comment here that this will log to screen on failure before returing?
Also, now this can be made a normal comment since it is not public anymore.
</blockquote></details>

Done. Since `protected` methods still appear in the Doxygen-generated website (see below). I think it will be best to keep it as a Doxygen comment.

![downcast-doxygen.png](https://files.reviewableusercontent.io/images/reviewable/1388098/RK_OEWBf3X6VWaosDMFCLunD/downcast-doxygen.png)

---

*[drake/multibody/joints/helical_joint.cc, line 29 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvQao9hKlADTBAI9NM:-KbvRAcoDzmO7W8WNSUc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/helical_joint.cc#L29)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.
</blockquote></details>

Done.

---

*[drake/multibody/joints/prismatic_joint.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvPbUBBIhAAjPLWY5Z:-KbvRGTEat_cSlyH0rpu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/prismatic_joint.cc#L27)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.
</blockquote></details>

Done.

---

*[drake/multibody/joints/revolute_joint.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-KbvPkxjf46Qz061SOGZ:-KbvRJJwpMoqeluZHSsY:b-yg1kce) ([raw file](https://github.com/robotlocomotion/drake/blob/1353929e5f0ba1e135c2c369c02136da40388361/drake/multibody/joints/revolute_joint.cc#L27)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW,  `if (downcasted_joint == nullptr) return false` is shorter.
</blockquote></details>

Done.

FTR, the reason I added braces was because of @jwnimmer-tri's comment here:

https://reviewable.io/reviews/robotlocomotion/drake/4775#-KaL0esYqP2PvSi1hNG0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4982#-:-KbvbudGJyhZHkP9dzUP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
